### PR TITLE
bpo-26467: Adds AsyncMock for asyncio Mock library support

### PR DESCRIPTION
**Pull request by lisroach**
This is my initial pass at supporting coroutine mocking via a new Mock subclass, CoroutineMock.

It can be used to mock out coroutines and have them validate as coroutines:

```python
>>> mock = CoroutineMock()
>>> asyncio.iscoroutinefunction(mock)
True
```

Test that a coroutine was awaited:

```python
class TestTests(unittest.TestCase):
    def test_assert_awaited(self):
        async def main():
            mock = CoroutineMock()
            with self.assertRaises(AssertionError):
                mock.assert_awaited()
            await mock()
            mock.assert_awaited()
        asyncio.run(main())
```
Also awaited_with, awaited_once_with, etc.

Things that I could use advice on:

1. Some of the code has been borrowed by asynctest (https://pypi.org/project/asynctest/) which is licensed under Apache 2. I have reached out to the author and am waiting to hear back, but am not sure how we feel about having Apache 2 licensed stuff in CPython. I could rewrite those spots if needed.

2. `inspect.iscoroutine` will return False for CoroutineMock objects. This is because the check is `isinstance(object, types.CoroutineType)`, and CoroutineTypes are concrete. I've looked into using something like, register, but it doesn't work here, so I need someway to make a Mock look like a types.CoroutineType but I am unsure how.

3. In CoroutineMockAssert tests, I have `asyncio.events._event_loop_policy` unset because it causes the env to change when running tests if it is not unset. There is probably a better way to do this where the environment doesn't get polluted?

4. I plan on working on getting more example test cases for the CoroutineArguments test section, would be happy for advice though.

And I will be writing up the documentation once the code has settled.

https://bugs.python.org/issue26467

### CODE REVIEW

**Comment by lisroach** [comment]
Talking with @asvetlov, it might be easier/more useful to implement this purely as an AwaitableMock, instead of mocking Coroutines themselves. This should also be more flexible in terms of inspect, because I could assert isawaitable() more easily than iscoroutine().

We also chatted about ideas around mocking AsyncIterators so we can continue to support __aiter__ and __anext__, but that will probably come in a separate commit.

**Comment by 1st1** [comment]
> Talking with @asvetlov, it might be easier/more useful to implement this purely as an AwaitableMock, instead of mocking Coroutines themselves.

+1.

**Comment by cjw296** [comment]
@lisroach - I'm afraid I don't any anywhere near enough experience with asyncio to sensibly review this, however one request: please can you split any tests that use the 3.7+ async syntax into their own file (`test..._async.py` or something) so that this can more easily be merged into the rolling backport.

I'm not sure if this will be possible, but if you could avoid using any 3.7+ syntax inside mock.py itself, that would, again, make backporting significantly easier :-)

**Comment by lisroach** [comment]
@cjw296 Good point about the backporting. So separating the tests is no problem, easy-peasy. For the internals of mock.py itself it's a bit more challenging because of the `async def __anext__` I need.

What if I try to wrap all the AsyncMock related code in mock.py in some regexable comment block (like `# Nonbackportable AsyncMock`), and all code within those blocks could be searched for and removed before backporting?

**Comment by cjw296** [comment]
What might work better is if you could put the `SyncMock` class and anything else that needs 3.7+ syntax into a `mockasync.py` next to `mock.py` and then do a `from .mockasrync import *` inside `mock.py`. That way, the code can stay in one place and we can make that a conditional import in the backport.

**Comment by lisroach** [comment]
I don't think it's possible to separate it out into a different file, the code is too integrated with the mock code itself we can't avoid the circular dependencies. I've moved all of the 3.7+ syntax code to the bottom of mock.py so it's all in one spot, will it be possible for it to be dropped when backporting with it like this?

**Comment by asvetlov** [comment]
@cjw296 the PR not only adds `AsyncMock` class but hacks already existing mock classes.
There is nothing new, mock is already has circular dependencies.

Please elaborate how `from .mockasync import *` at the end for `mock.py` can make backporting easier?
Do you want to remove this line and all usages for `mockasync` objects from `mock.py` during a backport?

File: Lib/unittest/mock.py
Comment (tirkarthi) [inline Lib/unittest/mock.py]:
```suggestion
import asyncio
```

File: Lib/unittest/mock.py
Comment (cjw296) [inline Lib/unittest/mock.py:922]: How much of all this kind of thing could be avoided if the docs just said "if you want to use async stuff, even to spec, just use `AsyncMock` from the start?"

**Comment by cjw296** [comment]
@asvetlov - yeah, I've commented on the "hacks already existing mock classes". This doesn't feel like a good idea to me, do we really need it?

The elaboration requested, would be this in cpython:

```
mock.py:
...all existing mock stuff...
from .mockasync import *

mockasync.py:
from .mock import *  # should work as we've already got the unittest.mock setup above
... anything using 3.7+ syntax...
```

This would be backported as:

```
mock.py:
...all existing mock stuff...
if sys.version_info >= (3, 7, ):
    from .mockasync import *
```
...so the patch can stay pretty much identical bar adding one line and changing another. It does require all Python 3.7+-only syntax going into mockasync.py.

File: Lib/unittest/mock.py
Comment (lisroach) [reply Lib/unittest/mock.py:922]: Some amount of integration could be avoided, although not all (these lines you commented on for example, would still stay to make child_mocks work). 

It could make mostly the __new__ less complicated. I'm not sure if I'd like the loss of functionality though

**Comment by lisroach** [comment]
@voidspace would you mind taking a look?

If the speccing stuff is adding too much complication for backporting we could pull it out. That would pull out most of the integrated code, although not all because child_mocks is still integrated. I don't think we want to lose the child_mock functionality, maybe there is a workaround somewhere there.

**Comment by voidspace** [comment]
It looks like to me that it's not possible to pull out all the code incompatible with Python 2.7 into another file due to the necessary child mock integration. My feeling is that we should have the best possible asyncio integration in Python 3.8, not sacrificing functionality for the sake of the backport, and deal with the issues that raises in the backport.

It may mean that this functionality has to be pulled, or reduced, in the backport unfortunately.

**Comment by cjw296** [comment]
@voidspace - it's not 2.7 that worries me so much as 3.6 (which also include pypy3), 3.5 and 3.4. 3.6 is a biggie as plenty of things (RHEL 8, for example) are pinned to that...

I'm a little confused as to why the child mock integration can't be kept to the `MockAsyncio` subclass?

File: Lib/unittest/mock.py
Comment (voidspace) [inline Lib/unittest/mock.py]: if not issubclass(cls, AsyncMock) ?

File: Lib/unittest/mock.py
Comment (voidspace) [reply Lib/unittest/mock.py:922]: It would mean speccing an object that has async attributes/methods etc wouldn't produce async compatible mocks where it should.

**Review by voidspace — COMMENTED** [review]
> @voidspace - it's not 2.7 that worries me so much as 3.6 (which also include pypy3), 3.5 and 3.4. 3.6 is a biggie as plenty of things (RHEL 8, for example) are pinned to that...
> 
> I'm a little confused as to why the child mock integration can't be kept to the `MockAsyncio` subclass?

The child mock support means that when using speccing you get AsyncMocks where appropriate. Pulling that out would be a loss of functionality. Full integration into Mock objects is the reason this has to be done in mock in the first place, rather than just as a third party extension to mock.

**Comment by asvetlov** [comment]
`async/await` keywords are available starting from Python 3.5

So technically backport to 3.5, 3.6 and 3.7 *just works* without any code modification. We didn't test it explicitly against different python versions though but even something doesn't work on Python 3.5 the fix should be trivial.

Support for 3.4 is doable but requires much more code (handling ugly `yield from` and `@asyncio.coroutine` instead of `async`/`await` builtin keywords).

I'm curious does mock backport for Python 3.4 makes sense since 3.4 end-of-life reached?

**Comment by cjw296** [comment]
Ah, for some reason I thought the new syntax only arrived in 3.7. 
I've got no great love for 3.4, and it might make sense to land this patch into the backport after we've decided to drop support for 2.7 and 3.4, in that case. @voidspace, thoughts?

**Comment by voidspace** [comment]
> Ah, for some reason I thought the new syntax only arrived in 3.7.
> I've got no great love for 3.4, and it might make sense to land this patch into the backport after we've decided to drop support for 2.7 and 3.4, in that case. @voidspace, thoughts?

That sounds good to me.

File: Lib/unittest/mock.py
Comment (lisroach) [reply Lib/unittest/mock.py]: Good catch, that is better.

File: Lib/unittest/test/testmock/testasync.py
Comment (asvetlov) [inline Lib/unittest/test/testmock/testasync.py]: Is this required?
AsyncMock doesn't change the asyncio loop/policy state anymore

File: Lib/unittest/test/testmock/testasync.py
Comment (asvetlov) [inline Lib/unittest/test/testmock/testasync.py]: I believe this code is not necessary

File: Lib/unittest/test/testmock/testasync.py
Comment (asvetlov) [inline Lib/unittest/test/testmock/testasync.py]: Policy again

File: Lib/unittest/test/testmock/testasync.py
Comment (asvetlov) [inline Lib/unittest/test/testmock/testasync.py]: Policy?

File: Lib/unittest/test/testmock/testasync.py
Comment (asvetlov) [inline Lib/unittest/test/testmock/testasync.py]: Policy?

File: Lib/unittest/test/testmock/testasync.py
Comment (asvetlov) [inline Lib/unittest/test/testmock/testasync.py]: Policy?

File: Lib/unittest/test/testmock/testasync.py
Comment (lisroach) [reply Lib/unittest/test/testmock/testasync.py]: So if I run the tests without this, I get an error that I am altering the execution environment:
```
1 test altered the execution environment:
    test_unittest
```
I thought it was happening in _AwaitEvent? But I am not sure.

Comment (asvetlov) [reply Lib/unittest/test/testmock/testasync.py]: Maybe that's because `asyncio.run()` alters the environment by definition.
I'll take a look.

Comment (asvetlov) [reply Lib/unittest/test/testmock/testasync.py]: I figured out the issue.
Before any asyncio calls, the default event loop policy is `None`.
To instantiate a loop you need a policy.
The policy is created on demand.
So, basically after any loop creation, the policy is not `None` but something concrete like `_UnixDefaultEventLoopPolicy`.

Cleaning it up at the end of test is correct.
To simplify the code you can add simple module teardown function: 
```
def tearDownModule():
    asyncio.set_event_loop_policy(None)
```
asyncio tests do this, I've missed that fact.

**Review by asvetlov — APPROVED** [review]
Overall looks good.
Please fix documentation build error.

Using `tearDownModule()` for cleaning up asyncio event loop policy in one place makes sense too.

**Comment by asvetlov** [comment]
Would be nice to land the PR on Python 3.8 (along with #13228 for `unittest.AsyncioTestCase`)

**Comment by cjw296** [comment]
Of interest for the backpor: does pytest yet have support for async test cases?

**Comment by asvetlov** [comment]
It supports async tests by plugins: `pytest-asyncio` for asyncio, `pytest-trio` for trio, `pytest-tornasync` for Tornado etc.
pytest plugin system allows integrating async tests smoothly (almost :) ) without hacking the core itself.

**Comment by cjw296** [comment]
Have you verified those are actually what they appear to be? I attempted to use `pytest-twisted` but found it to be naive and missing most of the important stuff so had to go back to trial...

**Comment by asvetlov** [comment]
I use `pytest-asyncio` at my job intensively.
Works ok, while I have to fix a couple of things after Python 3.8 beta freeze when I'll have free time.
Pretty sure that @njsmith keeps `pytest-trio` in a good shape as well.

File: Lib/unittest/test/testmock/testasync.py
Comment (tirkarthi) [inline Lib/unittest/test/testmock/testasync.py]: Nit: Some of the tests use `self.assertIsInstance` and some use `self.assertTrue(isinstance(obj, AsyncMock))` . I don't see `isinstance` used outside of testasync and using `self.assertIsInstance` would help in better error messages

File: Lib/unittest/mock.py
Comment (tirkarthi) [inline Lib/unittest/mock.py:2481]: Any reason behind this restriction since I don't see this for normal functions. Please add a test for this.

```python
def test_create_autospec_instance(self):
    with self.assertRaises(RuntimeError):
        create_autospec(async_func, instance=True)
```

File: Lib/unittest/mock.py
Comment (tirkarthi) [inline Lib/unittest/mock.py:2506]: Could this be documented as an attribute? I can see a reference in `reset_mock` but it would be good to have it in the docs like `call_count` is documented as a public attribute. Similar request for `await_args_list`

File: Lib/unittest/mock.py
Comment (tirkarthi) [inline Lib/unittest/mock.py]: I think `_is_coroutine` should be `asyncio.coroutines._is_coroutine`. Personally I think the attribute setup logic could be moved out of `create_autospec` as a utility function and would also help in backport where the setup function could just be commented.

```python
def test_create_autospec(self):
    spec = create_autospec(async_func)
```

```
======================================================================
ERROR: test_create_autospec (Lib.unittest.test.testmock.testasync.AsyncAutospecTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/test/testmock/testasync.py", line 156, in test_create_autospec
    spec = create_autospec(async_func)
  File "/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/unittest/mock.py", line 2504, in create_autospec
    mock._is_coroutine = _is_coroutine
NameError: name '_is_coroutine' is not defined
```

----------------------------------------------------------------------

**Review by tirkarthi — CHANGES_REQUESTED** [review]
I think it would be good to have `create_autospec(async_func)` fixed which would be a oneliner. Minor nits and clarifications. More tests and docs could be added after it's merged for beta 1. Please also add a WhatsNew entry since this is great to have and could also help with asyncio test cases in stdlib. Thanks @lisroach .

**Comment by tirkarthi** [comment]
@mariocj89 It would be great if you are available for a review of this. Thanks.

**Comment by tirkarthi** [comment]
`ResourceWarning` in `testasync.AsyncMockTest.test_future_isfuture`

```bash
➜  cpython git:(pr_9296) ✗ ./python.exe -m unittest -vv Lib.unittest.test.testmock.testasync.AsyncMockTest.test_future_isfuture
test_future_isfuture (Lib.unittest.test.testmock.testasync.AsyncMockTest) ... ok

----------------------------------------------------------------------
Ran 1 test in 0.036s

OK
<_UnixSelectorEventLoop running=False closed=False debug=False>
/Users/karthikeyansingaravelan/stuff/python/cpython/Lib/asyncio/base_events.py:653: ResourceWarning: unclosed event loop <_UnixSelectorEventLoop running=False closed=False debug=False>
  _warn(f"unclosed event loop {self!r}", ResourceWarning, source=self)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
```

**Review by 1st1 — APPROVED** [review]
@lisroach this is a huge piece of work, thank you for doing this.

IMO the PR is ready to be merged.  The proposed API is great (and is in line with the existing "sync" API), the new docs are on point, and the implementation looks polished enough (at least I don't see anything that needs to be fixed).

**Comment by 1st1** [comment]
@lisroach It looks like there are a couple of bugs in docs markups (see the CI).  Please fix those and hit the merge button! :)

File: Lib/unittest/mock.py
Comment (1st1) [reply Lib/unittest/mock.py:2506]: +1.

**Comment by 1st1** [comment]
@tirkarthi Most of your comments are on point, but I think we can address them once this is merged.

**Comment by tirkarthi** [comment]
@1st1 I would request create_autospec to be fixed as it's one line change causing NameError before beta release and if possible as part of this PR. Fixing the ResourceWarning would also be great just to make sure there is no buildbot failure outside of primary CI since I don't know the configuration in which they are run. 

More tests and docs could be added after beta and I would be happy to help with follow up PRs for the same. Thanks for approving this.

**Comment by tirkarthi** [comment]
@lisroach  Attached is a patch fixing docs and doctest that would make CI green. I tested it on my fork and seems to work fine. 

```patch
diff --git a/Doc/library/unittest.mock.rst b/Doc/library/unittest.mock.rst
index a8d05b216c..5d4737b60f 100644
--- a/Doc/library/unittest.mock.rst
+++ b/Doc/library/unittest.mock.rst
@@ -201,9 +201,11 @@ The Mock Class

 .. testsetup::

+    import asyncio
+    import inspect
     import unittest
     from unittest.mock import sentinel, DEFAULT, ANY
-    from unittest.mock import patch, call, Mock, MagicMock, PropertyMock
+    from unittest.mock import patch, call, Mock, MagicMock, PropertyMock, AsyncMock
     from unittest.mock import mock_open

 :class:`Mock` is a flexible mock object intended to replace the use of stubs and
@@ -885,9 +887,9 @@ object::
     ...
     >>> mock = MagicMock(async_func)
     >>> mock
-    <MagicMock spec='function' id='4568403696'>
+    <MagicMock spec='function' id='...'>
     >>> mock()
-    <coroutine object AsyncMockMixin._mock_call at 0x1104cb440>
+    <coroutine object AsyncMockMixin._mock_call at ...>

   .. method:: assert_awaited()

@@ -976,11 +978,11 @@ object::
       Assert the mock has been awaited with the specified calls.
       The :attr:`await_args_list` list is checked for the awaits.

-      If `any_order` is False (the default) then the awaits must be
+      If *any_order* is False (the default) then the awaits must be
       sequential. There can be extra calls before or after the
       specified awaits.

-      If `any_order` is True then the awaits can be in any order, but
+      If *any_order* is True then the awaits can be in any order, but
       they must all appear in :attr:`await_args_list`.

         >>> mock = AsyncMock()
```

File: Lib/unittest/test/testmock/testasync.py
Comment (lisroach) [reply Lib/unittest/test/testmock/testasync.py]: Thanks for digging into this @asvetlov! That makes sense. I'll update to use tearDownModule().

File: Lib/unittest/mock.py
Comment (lisroach) [reply Lib/unittest/mock.py]: I think you are right, this looks like a mistake, thanks for catching! I agree the attribute setup can be moved out into a utility function, let's do it in another PR.

File: Lib/unittest/test/testmock/testasync.py
Comment (tirkarthi) [inline Lib/unittest/test/testmock/testasync.py:141]: This test fails since the `RuntimeError` check is commented out.

File: Doc/library/unittest.mock.rst
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst]:
```suggestion
      ...     await mock(*args)
```

File: Doc/library/unittest.mock.rst
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst]:
```suggestion
      ...     await mock(*args)
```

File: Lib/unittest/test/testmock/testasync.py
Comment (lisroach) [reply Lib/unittest/test/testmock/testasync.py:141]: Ah yeah, sorry I was trying to remember the answer to your question on why it was needed. It might not be and I'm debating about taking it out, I think I don't fully understand the purpose of instance=True and what we would want the behavior to be if it is True with create_autospec on a async function.

File: Doc/library/unittest.mock.rst
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst]:
```suggestion
      >>> mock.await_args_list
```

File: Lib/unittest/test/testmock/testasync.py
Comment (tirkarthi) [reply Lib/unittest/test/testmock/testasync.py:141]: No problem, we can remove the restriction later if needed. I have added one more doctest suggestion that would make the CI green and this can be merged.

**Review by tirkarthi — APPROVED** [review]
LGTM. Buildbots are green and it will be great to have this merged for beta. I will create a followup PR that can be reviewed separately without holding this up any further. 

Thank you very much @lisroach :)

**Comment by lisroach** [comment]
Thanks for all the help and reviews everyone!

**Comment by mariocj89** [comment]
I know I am late to the party. Real apologies about it, I had LASIK past week and could not check this on the weekend. I don't know how many times people have asked about `AsyncMock`, this is just great :).

My main comment here would be whether it is useful or confusing the fact that `AsyncMock` will generate "inner mocks" when attributes are called on it. I would find that confusing though I am far from an expert on async. Is there any use-case were calling an attribute in an async object it yields naturally another async object?

Example, not sure this is expected:

```python
>>> from unittest.mock import AsyncMock
>>> m = AsyncMock()
>>> m.a.return_value = 1
>>> print(m.a)
<AsyncMock name='mock.a' id='140389785043824'>
>>> print(m.a())
<coroutine object AsyncMockMixin._mock_call at 0x7faf0ab8d240>
<stdin>:1: RuntimeWarning: coroutine 'AsyncMockMixin._mock_call' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

I think I'd use the `AsyncMock` to mock specific function calls. It is also really handy that it works with patch. I was just wondering whether getting an attribute of the `AsyncMock` should return:

1. Nothing at all, attribute error and let users set attributes if desired.
2. A normal mock as that might be more common (Patching an async callable class for example).
3. Default to async mock as the current implementation does. People can set normal mocks if wanted.

If this was thought and 3 was chosen purposely because it makes more sense for async users, all good (as said, not an expert in async at all), just raising the question.

**Comment by mariocj89** [comment]
Example, how do we expect users to patch the following class?

```python
class X:
  async def a(self):
    return "a"
  def b(self):
    return "b"
```

If the plan is to use AsyncMock for it, the following does not work:
```python
m = AsyncMock(**{"a.return_value": "mocka", "b.return_value": "mockb"})
async def check(m):
    print(m.b())
    print(await m.a())
```

Output:
```
>>> asyncio.run(check(m))
<coroutine object AsyncMockMixin._mock_call at 0x7f39dea8cac0>
<stdin>:2: RuntimeWarning: coroutine 'AsyncMockMixin._mock_call' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
mocka
```

If we say that users should specify the exact type of mock on each of the attribute, what is the purpose of automatically generating them for `AsyncMock` then?

As said, minor concern, there might be no better solution, but wanted to have a look at it.

**Comment by asvetlov** [comment]
Thanks for the comment.
I see a bunch of `RuntimeWarning: coroutine 'AsyncMockMixin._mock_call' was never awaited` when running `tests.test_asyncio` suite.
Need to investigate.
See https://bugs.python.org/issue37015 for the reference.

**Comment by lisroach** [comment]
Thanks for pointing that out @mariocj89! I made it recursive asyncmocks more as a way to copy the normal functionality of Mock, which creates an internal mock of the same kind when calling an attribute, for example:
```
>>> from unittest.mock import MagicMock
>>> m = MagicMock()
>>> m.a.return_value = 1
>>> print(m.a)
<MagicMock name='mock.a' id='4490221040'>
```

It is mentioned in the docstring "By default child mocks will be the same type as the parent", although this can be overridden. 
I wanted to mimic the same functionality because I do think that is what is expected when using the mock library. 

The current functionality is that if you want to mock a mixed class, you use a normal mock and it figures out for you which functions are async:

`
>>> class X:
...   async def a(self):
...     return "a"
...   def b(self):
...     return "b"
...
>>> m = MagicMock(X, autospec=True)
>>> m.b
<MagicMock name='mock.b' id='4490513224'>
>>> m.a
<AsyncMock name='mock.a' id='4490514088'>
`

But you are right if someone was attempting to build their own class via mock the would need to specify the mock type of the arguments. I don't usually use mock this way so I hadn't considered that use case. In your example it looks like there would be no way to tell it is supposed to be an awaitable mock until the actual call occurs, which might be tricky to implement. If you think it will be useful though I don't mind if you open up a new bug and we can figure it out there!

**Comment by jeremycline** [comment]
Hi folks,

I've hit an [issue](https://bugzilla.redhat.com/show_bug.cgi?id=1717665#c5) I think is related to this PR in Python 3.8b1. The minimal reproducer I came up with is:

```
from unittest import mock

mock_obj = mock.MagicMock()
mock_obj.mock_func = mock.MagicMock(spec=lambda x: x)

with mock.patch.object(mock_obj, "mock_func") as nested:
    print(type(nested))
```

What happens is that ``nested`` is an AsyncMock in Python 3.8, but a MagicMock in Python 3.7. The reason is when the MagicMock has a function spec, ``__code__`` is defined so the check for the coroutine flags returns a truthy MagicMock object. In other words, ``import inspect; from unittest import mock; inspect.iscoroutinefunction(mock.MagicMock(spec=lambda x: x))`` is ``True`` (and that holds in Python 3.7 as well).

I thought about just submitting a fix, but it's not clear to me whether adding a check to ``unittest.mock._is_async_obj`` for MagicMock and behaving differently is truly the right fix. Having ``asyncio.iscoroutinefunction`` return ``True`` seems wrong for MagicMocks with function specs. Maybe MagicMock needs to have something to make it behave correctly there?

**Comment by lisroach** [comment]
Hey @jeremycline, would you mind making a bug on bugs.python.org? It will be best to centralize discussion there. Thanks for reporting this! It is definitely a bug.

I understand the desire to change MagicMock to avoid this behavior, but I think it is the wrong approach. It is not evaluating to True because of the `__code__.co_flag` attribute, that doesn't get set until after the check for `_is_async_obj` has already evaluated truthy. 

The issue comes in when checking:

`bool(f.__code__.co_flags & flag)`

in `inspect._has_code_flag`. The `&` is a call, which with MagicMocks will result in another MagicMock object:

`<MagicMock name='mock.mock_func.__code__.co_flags.__and__()' id='4467594096'>`

It does not take into account the `flag` argument at all, and calling `bool()` on the MagicMock evaluates to True. 

The only way I see us getting around this would be to set the `mock.__code__.co_flags` attribute everytime a spec with `__code__` is passed in to MagicMock, but this changes the way mock typically behaves. It would be unexpected for one attribute to be set (I think) if others are not. 

It might be best to make a workaround in the `_is_async_obj`, although it feels more hacky.

**Comment by njsmith** [comment]
Maybe `bool(MagicMock)` should raise an error?

**Comment by tirkarthi** [comment]
> It might be best to make a workaround in the _is_async_obj, although it feels more hacky.

I feel this is a reasonable trade off that maybe we could check if `__code__` is actually a `CodeType` to ensure `MagicMock` is not passed to `iscoroutinefunction` to give a false positive. 

```
def _is_async_obj(obj):
    code = getattr(obj, '__code__', None)
    if code and isinstance(code, CodeType):
        return asyncio.iscoroutinefunction(obj) or inspect.isawaitable(obj)
    else:
        return False
```

**Comment by jeremycline** [comment]
> Hey @jeremycline, would you mind making a bug on bugs.python.org? It will be best to centralize discussion there. Thanks for reporting this! It is definitely a bug.
> 

Thanks for the quick response, I've filed an issue at https://bugs.python.org/issue37251.

File: Lib/unittest/mock.py
Comment (f0k) [inline Lib/unittest/mock.py:458]: Late note, not sure if anyone gets a notification for this: This broke a use case for me. Before this change, the `Mock` constructor did not call `getattr` on any of the attributes of the mocked object specification. I used it to mock an instance of a base class that does not have all its properties implemented, using the `**kwargs` in the `Mock` constructor to ensure that unimplemented properties that were accessed by any tests got a sane value. I can solve this by using a subclass that does not have unimplemented properties. But maybe it would be wise to skip all attributes here that appear in `**kwargs`, because they will be subsequently overwritten?

Comment (arhadthedev) [reply Lib/unittest/mock.py:458]: > not sure if anyone gets a notification for this

For this reason it's worth creating a new issue and mentioning this PR there.

Comment (cjw296) [reply Lib/unittest/mock.py:458]: @f0k - better yet, if you could work up a PR with a test that shows the problem you encountered, along with a fix for it, that would be better!

Comment (cjw296) [reply Lib/unittest/mock.py:458]: @f0k - better yet, if you could work up a PR with a test that shows the problem you encountered, along with a fix for it, that would be better!

Comment (tirkarthi) [reply Lib/unittest/mock.py:458]: @f0k some related issues . Please use search before creating the issue so that you can check for previous issues and make sure it's not duplicate. Thanks.

https://github.com/python/cpython/issues/89917
https://github.com/python/cpython/issues/85934

Comment (fok) [reply Lib/unittest/mock.py:458]: @tirkarthi you probably meant @f0k, not me

Comment (f0k) [reply Lib/unittest/mock.py:458]: > For this reason it's worth creating a new issue

I wasn't sure yet whether this warrants a fix, and was aiming low :)

> some related issues

Good spot, yes, they're both the same problem, and there are already three PRs proposing a fix. I will continue the discussion in #85934, as it's the older one of the two issues.

**Pull request merged**
